File: hw/pastor/u3/hw29Singleton/PYTHON/Controller/StockAlertController.py
from pymongo import MongoClient
from Model.Product import Product
from Config.AlarmConfig import AlarmConfig

class StockAlertController:

    def __init__(self):
        self.client = MongoClient("mongodb+srv://Mathews:Mathews2007@cluster0.6l9ibfh.mongodb.net/?appName=Cluster0")
        self.db = self.client["SingletonDB"]
        self.productsCollection = self.db["products"]

        self.config = AlarmConfig.getInstance()
        self.view = None

    def setView(self, view):
        self.view = view

    def checkStock(self):
        minimumStock = self.config.getMinimumStock()
        lowStockProducts = []

        for doc in self.productsCollection.find():
            product = Product(doc["_id"], doc["name"], doc["stock"])
            if product.stock <= minimumStock:
                lowStockProducts.append(product)

        self.view.showLowStockProducts(lowStockProducts)

    def updateMinimumStock(self, newValue: int):
        self.config.updateMinimumStock(newValue)
        self.view.showInfoMessage(f"Minimum stock updated to {newValue}")

File: hw/medina/u3/hw29singleton/singletonInJavaScript/Controller/StockAlertController.js
import { MongoClient } from "mongodb";
import { Product } from "../Model/Product.js";
import { AlarmConfig } from "../Config/AlarmConfig.js";

export class StockAlertController {
    constructor() {
        this.client = new MongoClient("mongodb+srv://Joseph:Joseph1751774793@cluster0.h8pi0ir.mongodb.net/?appName=Cluster0");
    }

    async init() {
        await this.client.connect();
        this.db = this.client.db("SingletonDB");
        this.productsCollection = this.db.collection("products");
        this.config = await AlarmConfig.getInstance();
    }

    async checkStock() {
        const min = this.config.getMinimumStock();
        const cursor = this.productsCollection.find();

        const lowStockProducts = [];

        for await (const doc of cursor) {
            const product = new Product(doc._id, doc.name, doc.stock);
            if (product.stock <= min) {
                lowStockProducts.push(product);
            }
        }

        return lowStockProducts;
    }

    async updateMinimumStock(newValue) {
        await this.config.updateMinimumStock(newValue);
    }
}
